File: Javascript/2.Settembre/43.DigitsVsLetters-22Settembre.js
//todo Data una stringa, restituisce "cifre" se la stringa ha più cifre che lettere, "lettere" se ha più lettere che cifre e "pareggio" se ha lo stesso numero di cifre e lettere.

//! Le cifre sono composte da 0 a 9.
//! Le lettere sono composte da a-z, maiuscole o minuscole.
//! Ignora qualsiasi altro carattere.

//? Tempo di esecuzione:
//* 6 minuti e 35 secondi

//?Codice scritto senza utilizzo di ChatGpt

function digitsOrLetters(str) {
  str.toLowerCase()
  let contatoreLet = 0;
  let contatoreNum = 0;

  let letters = "abcdefghilmnopqrstuvzywkjABCDEFGHILMNOPQRSTUVZWKYJ"
  let numbers = "0123456789"

  for(let i = 0; i < str.length; i++) {
    if(letters.includes(str[i])) contatoreLet++
    if(numbers.includes(str[i])) contatoreNum++
  }

  if(contatoreLet > contatoreNum) return "letters"
  if(contatoreLet < contatoreNum) return "digits"
  else return "tie"

}

console.log(digitsOrLetters("abc123!@#DEF"))

//?Codice scritto da ChatGpt
//* Non usata

//* ✅ Pro

//* Logica corretta

//* Ignora correttamente i caratteri speciali

//* Gestione chiara dei tre casi (letters / digits / tie)

//* Ciclo semplice e comprensibile

//! ❌ Contro

//! str.toLowerCase() non ha effetto (manca assegnazione)

//! Stringa letters incompleta e disordinata

//! Doppio controllo includes poco efficiente

//! Codice più complesso del necessario

//? Versione Chat Gpt 

function digitsOrLetters(str) {
  let letters = 0;
  let digits = 0;

  for (const char of str) {
    if (char >= "a" && char <= "z" || char >= "A" && char <= "Z") letters++;
    else if (char >= "0" && char <= "9") digits++;
  }

  if (letters > digits) return "letters";
  if (digits > letters) return "digits";
  return "tie";
}

//todo Considerazioni: 

//! N.B.

File: Python/2.Settembre/43.DigitsVsLetters-22Settembre.py
# =========================================================
# FILE TEMPLATE - ANALISI ESERCIZI PYTHON
# =========================================================

# //Prompt ChatGpt: questo è il modello. Invierò consegna esercizio e codice, e vorrei feedback strutturato

# =========================================================
# //todo Consegna dell'esercizio
# =========================================================

# Data una stringa, restituire:
# - "cifre" se ci sono più numeri che lettere
# - "lettere" se ci sono più lettere che numeri
# - "pareggio" se uguali
#
# Regole:
# - cifre: 0–9
# - lettere: a–z (maiuscole e minuscole)
# - altri caratteri ignorati

# =========================================================
# //* Codice scritto senza utilizzo di ChatGpt
# =========================================================

import re

def digits_or_letters(s):
    char = 0;
    numbers = 0;

    for value in s:
        if re.match("[a-zA-Z]", value):
            char += 1
        elif re.match("[0-9]", value):
            numbers += 1

    if char > numbers:
        return "letters";
    elif char < numbers:
        return "digits";
    else:
        return "tie"
        
print(digits_or_letters("abc123"))

# =========================================================
# //* Codice scritto da ChatGpt (versione migliorata)
# =========================================================

def digits_or_letters(s):
    letters = sum(c.isalpha() for c in s)
    digits = sum(c.isdigit() for c in s)

    if digits > letters:
        return "cifre"
    elif letters > digits:
        return "lettere"
    else:
        return "pareggio"

# =========================================================
# //* Pro e contro del mio codice
# =========================================================

# //* Pro
# - Logica corretta nel conteggio.
# - Ignora automaticamente i caratteri non validi.
# - Struttura semplice e comprensibile.
# - Funziona per lettere e numeri come richiesto.

# //! Contro
# - Uso di regex non necessario per un singolo carattere.
# - `re.match` è più lento e meno idiomatico di `isalpha()` / `isdigit()`.
# - Output non conforme alla consegna:
#   - "letters" invece di "lettere"
#   - "digits" invece di "cifre"
#   - "tie" invece di "pareggio"
# - Nomi variabili poco descrittivi (`char`, `numbers`).
# - Punteggiatura inutile nei return (`;`).

# =========================================================
# //* Consigli
# =========================================================

# - Il problema principale qui è la NON aderenza alla consegna:
#   i valori di ritorno devono essere ESATTAMENTE quelli richiesti.
#
# - In Python, meglio evitare regex per controlli così semplici:
#       c.isalpha()
#       c.isdigit()
#
# - Piccolo miglioramento mentale utile:
#   prima controlla SEMPRE l’output richiesto, poi scrivi il codice.
#
# - La tua logica è corretta, quindi il problema è solo di “traduzione
#   della consegna”, non di algoritmo.

# Valutazione: 8/10
#
# Buona logica, ma penalizzato per mismatch nei valori di output e uso regex non necessario.